lib/repl.js

```diff
@@ -258,21 +264,84 @@ function REPLServer(prompt,
       sawSIGINT = false;
     }
 
+    self._currentStringLiteral = null;
     self.bufferedCommand = '';
     self.lines.level = [];
     self.displayPrompt();
   });
 
+  function parseForStringLiteral(line) {
+    var previous = null, current = null;
+
+    for (var i = 0; i < line.length; i += 1) {
+      if (previous === '\\') {
+        // if it is a valid escaping, then skip processing
+        previous = current;
+        continue;
+      }
+
+      current = line.charAt(i);
+      if (current === self._currentStringLiteral) {
+        self._currentStringLiteral = null;
+      } else if (current === '\'' ||
+                 current === '"' &&
+                 self._currentStringLiteral === null) {
+        self._currentStringLiteral = current;
+      }
+      previous = current;
+    }
+  }
+
+  function getFinisherFunction(cmd, defaultFn) {
+    if ((self._currentStringLiteral === null &&
+         cmd.charAt(cmd.length - 1) === '\\') ||
+        (self._currentStringLiteral !== null &&
+         cmd.charAt(cmd.length - 1) !== '\\')) {
+
+      // If the line continuation is used outside string literal or if the
+      // string continuation happens with out line continuation, then fail hard
+      return function(e, ret) {
+        var error = e instanceof Recoverable ? e.err : e;
+        var args = [error];
+
+        if (arguments.length === 2) {
+          // adding second argument only if it is actually passed. Otherwise
+          // `undefined` will be printed when invalid REPL commands are used.
+          args.push(ret);
+        }
```

Comment:
This is still necessary because if we pass the second argument to the default function (which is finish) unless it is really passed to this function, the argument count will be two and the fiz for undefined will not work :(